src/Mobs/Path.cpp

```diff
@@ -0,0 +1,391 @@
+/* MCServer headers */
+#include "Globals.h"
+#include "../World.h"
+#include "../Chunk.h"
+#endif
+
+
+#include "Path.h"
+
+#include <stdio.h>
+#include <exception>
+#include <cmath>
+
+#define DISTANCE_MANHATTEN 0  // 1: More speed, a bit less accuracy 0: Max accuracy, less speed.
+#define HEURISTICS_ONLY 0  // 1: Much more speed, much less accurate.
+#define CALCULATIONS_PER_STEP 5  // Higher means more CPU load but faster path calculations.
+// The only version which guarantees the shortest path is 0, 0.
+
+enum class eCellStatus {OPENLIST,  CLOSEDLIST,  NOLIST};
+struct cPathCell
+{
+	Vector3d m_Location;   // Location of the cell in the world
+	int m_F, m_G, m_H;     // F, G, H as defined in regular A*.
+	eCellStatus m_Status;  // Which list is the cell in? Either non, open, or closed.
+	cPathCell * m_Parent;  // Cell's parent, as defined in regular A*.
+	bool m_IsSolid;	       // Is the cell an air or a solid? Partial solids are currently considered solids.
+};
+
+
+
+
+
+bool compareHeuristics::operator()(cPathCell * & a_V1, cPathCell * & a_V2)
+{
+	return a_V1->m_F > a_V2->m_F;
+}
+
+
+
+
+
+/* cPath implementation */
+#ifndef COMPILING_PATHFIND_DEBUGGER
+bool cPath::IsSolid(const Vector3d & a_Location)
+{
+	int ChunkX, ChunkZ;
+	m_Item_CurrentBlock = a_Location;
+	m_Item_SetMode = false;  // Causes item() to tell us whether the block is solid or not, this field is for testing and will be removed later.
+	// printf("IsSolid called: (%d %d %d)\n", (int)m_Item_CurrentBlock.x, (int)m_Item_CurrentBlock.y, (int)m_Item_CurrentBlock.z);
+	cChunkDef::BlockToChunk(a_Location.x, a_Location.z, ChunkX, ChunkZ);
+	return !m_World->DoWithChunk(ChunkX, ChunkZ, *this);
+}
+
+
+
+
+// Incomplete
+bool cPath::Item(cChunk * a_Chunk)  // returns FALSE if there's a solid or if we failed.
+{
+	int RelX = m_Item_CurrentBlock.x - a_Chunk->GetPosX() * cChunkDef::Width;
+	int RelZ = m_Item_CurrentBlock.z - a_Chunk->GetPosZ() * cChunkDef::Width;
+	
+	// If it's setmode, Sets a block to cobblestone and exits, will be removed eventually.
+	if (m_Item_SetMode)
+	{
+		a_Chunk->FastSetBlock(RelX, m_Item_CurrentBlock.y, RelZ, E_BLOCK_COBBLESTONE, 0);
+		return false;
+	}
+	
+	
+	if (!a_Chunk->IsValid())
+	{
+		// printf("cPath::item - Invalid chunk. Probably nobody is standing there. (%d %d %d)\n", (int)m_Item_CurrentBlock.x, (int)m_Item_CurrentBlock.y, (int)m_Item_CurrentBlock.z);
+		return false;
+	}
+	BLOCKTYPE BlockType;
+	NIBBLETYPE BlockMeta;
+	a_Chunk->GetBlockTypeMeta(RelX, m_Item_CurrentBlock.y, RelZ, BlockType, BlockMeta);
+	return (!cBlockInfo::IsSolid(BlockType));
+	
+	// TODO Maybe I should queue several blocks and call item() at once for all of them for better performance?
+	// I think Worktycho said each item() call needs 2 locks.
+	
+}
+#endif
+
+
+
+
+
+cPath::cPath(cWorld * a_World,
+const Vector3d & a_StartingPoint, const Vector3d & a_EndingPoint, int a_MaxSteps,
+double a_BoundingBoxWidth, double a_BoundingBoxHeight,
+int a_MaxUp, int a_MaxDown)
+{
```

Comment:
Add indent for the params, the closing parenthesis should be on a separate line (one less indent) and the first param should be on a separate line.